File: sdk/velma_recommender.py
# Copyright (c) 2024-present The Wayfinder Trust - Jonathan Luethke
# All Rights Reserved.
#
# This file is part of the Velma Ecosystem — SkillChain.
# Unauthorized copying, modification, distribution, or use of this file,
# via any medium, is strictly prohibited.
#
# Proprietary and Confidential.
# Classification: PROPRIETARY
"""
velma_recommender.py
====================

Velma knows what you need before you ask.

Observes your trainer state, skill history, streaks, time patterns,
and category gaps — then recommends the chain you should run RIGHT NOW.
No searching. No browsing. Velma just tells you.

Example::

    from skillchain.sdk.velma_recommender import VelmaRecommender

    velma = VelmaRecommender()
    rec = velma.what_now()
    print(rec)
    # → "You haven't checked in on yourself in 3 weeks. Run am-i-okay."
"""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass, field
from datetime import date, datetime, timedelta
from pathlib import Path
from typing import Any, Optional

from .gamification import GamificationEngine, SKILL_CATEGORIES, TrainerState

logger = logging.getLogger(__name__)

_MARKETPLACE_DIR = Path(__file__).resolve().parent.parent / "marketplace"


# ---------------------------------------------------------------------------
# Recommendation
# ---------------------------------------------------------------------------

@dataclass
class VelmaRecommendation:
    """What Velma thinks you should do right now."""
    chain_name: str
    reason: str
    priority: int  # 1 = most urgent, 5 = nice to have
    category: str
    nudge: str  # The human-friendly message


# ---------------------------------------------------------------------------
# Time rules
# ---------------------------------------------------------------------------

_DAY_OF_WEEK_CHAINS = {
    0: ("weekly-ops", "It's Monday. Time to plan your week."),
    6: ("weekly-ops", "Sunday night. Set up tomorrow."),
}

_TIME_OF_DAY_CHAINS = {
    # (start_hour, end_hour): (chain, nudge)
    (5, 8): ("morning-routine", "Early bird gets the plan."),
}


# ---------------------------------------------------------------------------
# Category health thresholds
# ---------------------------------------------------------------------------

_CATEGORY_NUDGES = {
    "Life": ("am-i-okay", "You haven't done a life check-in lately."),
    "Health": ("healthy-living-reset", "Your health category is untouched."),
    "Money": ("money-wake-up", "When's the last time you looked at your money?"),
    "Career": ("should-i-quit", "Your career chains have been quiet."),
    "Learning": ("learn-anything", "You haven't learned anything new in a while."),
    "Onboarding": ("tech-confidence-builder", "Some onboarding skills might help."),
}

# New user progression
_NEW_USER_SEQUENCE = [
    ("tech-confidence-builder", "Let's start with the basics.", 0),
    ("learn-anything", "Pick something you've always wanted to learn.", 3),
    ("am-i-okay", "Let's check in on how you're doing.", 7),
    ("weekly-ops", "Time to build a weekly rhythm.", 10),
    ("before-you-leap", "Got any big decisions brewing?", 14),
]


# ---------------------------------------------------------------------------
# Recommender
# ---------------------------------------------------------------------------

class VelmaRecommender:
    """Observes your state and recommends what to do next."""

    def __init__(self) -> None:
        self._engine = GamificationEngine()
        self._state = self._engine._state
        self._chains = self._load_chains()

    def what_now(self) -> list[VelmaRecommendation]:
        """What should you do right now? Returns ranked recommendations."""
        recs: list[VelmaRecommendation] = []

        # 1. New user check
        if self._state.total_skill_runs == 0:
            recs.append(VelmaRecommendation(
                chain_name="tech-confidence-builder",
                reason="first_time_user",
                priority=1,
                category="onboarding",
                nudge="Welcome. Let's start with understanding what AI can do for you.",
            ))
            return recs

        # 2. Streak about to die
        streak_rec = self._check_streak()
        if streak_rec:
            recs.append(streak_rec)

        # 3. Time-based recommendations
        time_rec = self._check_time()
        if time_rec:
            recs.append(time_rec)

        # 4. Category gaps
        gap_recs = self._check_category_gaps()
        recs.extend(gap_recs)

        # 5. Progression — what's the natural next step
        prog_rec = self._check_progression()
        if prog_rec:
            recs.append(prog_rec)

        # 6. Quest completion
        quest_rec = self._check_quests()
        if quest_rec:
            recs.append(quest_rec)

        # Sort by priority
        recs.sort(key=lambda r: r.priority)

        # Dedupe by chain name
        seen: set[str] = set()
        unique: list[VelmaRecommendation] = []
        for r in recs:
            if r.chain_name not in seen:
                seen.add(r.chain_name)
                unique.append(r)

        return unique[:5]  # Top 5

    def nudge(self) -> str:
        """Get a single human-friendly nudge message."""
        recs = self.what_now()
        if not recs:
            return "You're all caught up. Take a break or explore something new."
        return recs[0].nudge

    # -- Checks ------------------------------------------------------------

    def _check_streak(self) -> Optional[VelmaRecommendation]:
        """Is the streak about to break?"""
        if self._state.streak_current == 0:
            return None

        today = date.today().isoformat()
        if self._state.streak_last_date == today:
            return None  # Already ran today

        yesterday = date.fromordinal(date.today().toordinal() - 1).isoformat()
        if self._state.streak_last_date == yesterday:
            # Streak is alive but hasn't been maintained today
            return VelmaRecommendation(
                chain_name=self._quick_chain(),
                reason="streak_maintenance",
                priority=1,
                category="life",
                nudge=f"Your {self._state.streak_current}-day streak is on the line. Run anything to keep it alive.",
            )
        return None

    def _check_time(self) -> Optional[VelmaRecommendation]:
        """Time-of-day or day-of-week recommendations."""
        now = datetime.now()
        dow = now.weekday()

        # Day of week
        if dow in _DAY_OF_WEEK_CHAINS:
            chain, nudge = _DAY_OF_WEEK_CHAINS[dow]
            # Only suggest if they haven't run it today
            if chain not in self._state.daily_runs_today:
                return VelmaRecommendation(
                    chain_name=chain,
                    reason="day_of_week",
                    priority=2,
                    category="life",
                    nudge=nudge,
                )

        # Time of day
        hour = now.hour
        for (start, end), (chain, nudge) in _TIME_OF_DAY_CHAINS.items():
            if start <= hour < end and chain not in self._state.daily_runs_today:
                return VelmaRecommendation(
                    chain_name=chain,
                    reason="time_of_day",
                    priority=2,
                    category="life",
                    nudge=nudge,
                )

        return None

    def _check_category_gaps(self) -> list[VelmaRecommendation]:
        """Find categories the user has never explored."""
        discovered = set(self._state.skills_discovered)
        recs: list[VelmaRecommendation] = []

        for cat, skills in SKILL_CATEGORIES.items():
            found = [s for s in skills if s in discovered]
            if len(found) == 0 and cat in _CATEGORY_NUDGES:
                chain, nudge = _CATEGORY_NUDGES[cat]
                recs.append(VelmaRecommendation(
                    chain_name=chain,
                    reason=f"category_gap_{cat.lower()}",
                    priority=3,
                    category=cat.lower(),
                    nudge=nudge,
                ))

        return recs[:2]  # Max 2 category gap suggestions

    def _check_progression(self) -> Optional[VelmaRecommendation]:
        """What's the natural next step based on what they've done?"""
        completed = set(self._state.chains_completed)

        # If they just finished should-i-quit, suggest the next step
        if "should-i-quit" in completed and "job-search-blitz" not in completed:
            return VelmaRecommendation(
                chain_name="job-search-blitz",
                reason="progression",
                priority=3,
                category="career",
                nudge="You thought about quitting. Ready to start looking?",
            )

        if "job-search-blitz" in completed and "career-pivot" not in completed:
            return VelmaRecommendation(
                chain_name="career-pivot",
                reason="progression",
                priority=4,
                category="career",
                nudge="Job search is rolling. Want to explore a different field entirely?",
            )

        # Business progression
        if "micro-business-launch" in completed and "content-to-customers" not in completed:
            return VelmaRecommendation(
                chain_name="content-to-customers",
                reason="progression",
                priority=3,
                category="business",
                nudge="Business is launched. Time to get customers with content.",
            )

        if "startup-validation" in completed and "first-customers" not in completed:
            return VelmaRecommendation(
                chain_name="first-customers",
                reason="progression",
                priority=3,
                category="business",
                nudge="Idea is validated. Time to find your first 10 customers.",
            )

        # Life progression
        if "stuck-to-unstuck" in completed and "before-you-leap" not in completed:
            return VelmaRecommendation(
                chain_name="before-you-leap",
                reason="progression",
                priority=4,
                category="life",
                nudge="You broke the paralysis. Ready to make the big move?",
            )

        # Health progression
        if "healthy-living-reset" in completed and "morning-routine" not in completed:
            return VelmaRecommendation(
                chain_name="morning-routine",
                reason="progression",
                priority=4,
                category="life",
                nudge="Health plan is set. Build a morning routine to sustain it.",
            )

        return None

    def _check_quests(self) -> Optional[VelmaRecommendation]:
        """Suggest a chain that helps complete a daily quest."""
        quests = self._engine.get_daily_quests()
        for q in quests:
            if not q["completed"]:
                if q["type"] == "chain":
                    return VelmaRecommendation(
                        chain_name=self._quick_chain(),
                        reason="quest_completion",
                        priority=4,
                        category="life",
                        nudge=f"Daily quest: {q['text']} (+{q['xp']} XP)",
                    )
                if q["type"] == "new_skill":
                    # Find a category they haven't explored
                    discovered = set(self._state.skills_discovered)
                    for cat, skills in SKILL_CATEGORIES.items():
                        for s in skills:
                            if s not in discovered:
                                return VelmaRecommendation(
                                    chain_name=_CATEGORY_NUDGES.get(cat, ("learn-anything", ""))[0],
                                    reason="quest_new_skill",
                                    priority=4,
                                    category=cat.lower(),
                                    nudge=f"Try something new: {s.replace('-', ' ')} (+{q['xp']} XP)",
                                )
        return None

    # -- Helpers -----------------------------------------------------------

    def _quick_chain(self) -> str:
        """Return a short chain for streak maintenance."""
        # Prefer chains the user hasn't run yet
        completed = set(self._state.chains_completed)
        quick_options = [
            "morning-routine", "weekly-ops", "health-check-prep",
            "learn-anything", "meeting-to-action",
        ]
        for c in quick_options:
            if c not in completed:
                return c
        return "morning-routine"

    def _load_chains(self) -> list[dict]:
        """Load chain definitions."""
        chains_dir = _MARKETPLACE_DIR / "chains"
        chains = []
        if chains_dir.exists():
            for f in sorted(chains_dir.glob("*.chain.json")):
                try:
                    chains.append(json.loads(f.read_text(encoding="utf-8")))
                except Exception:
                    continue
        return chains

File: mcp-node/src/velma-recommender.ts
/**
 * Velma Recommender — "what now?" intelligent recommendations.
 * Port of sdk/velma_recommender.py
 */
import { readdirSync, readFileSync, existsSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import { GamificationEngine, SKILL_CATEGORIES } from "./gamification.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export interface VelmaRecommendation {
  chain_name: string;
  reason: string;
  priority: number;
  category: string;
  nudge: string;
}

const DAY_CHAINS: Record<number, [string, string]> = {
  0: ["weekly-review", "Sunday is for reflecting. Let's review your week."],
  1: ["weekly-ops", "Monday energy. Let's plan the week."],
  2: ["learn-anything", "Tuesday is learning day. Pick something new."],
  3: ["content-machine", "Midweek momentum. Create something."],
  4: ["networking-blitz", "Thursday is for people. Expand your network."],
  5: ["side-hustle-launch", "Friday build day. Ship something."],
  6: ["am-i-okay", "Saturday check-in. How are you really doing?"],
};

const TIME_CHAINS: Array<[number, number, string, string]> = [
  [5, 9, "morning-routine", "Early bird gets the worm. Let's start your day right."],
  [9, 12, "weekly-ops", "Prime focus hours. Let's get things done."],
  [12, 14, "learn-anything", "Lunch break learning. Feed your brain."],
  [14, 17, "content-machine", "Afternoon creative block. Make something."],
  [17, 20, "budget-builder", "Evening admin. Get your money right."],
  [20, 23, "daily-reflection", "Wind down. Reflect on today."],
];

const CATEGORY_NUDGES: Record<string, [string, string]> = {
  Life: ["am-i-okay", "You haven't explored Life skills yet. Start with a check-in."],
  Career: ["career-pivot", "Career skills untouched. Let's figure out your next move."],
  Business: ["side-hustle-launch", "Business skills waiting. Ready to build something?"],
  Health: ["burnout-recovery", "Health matters. Let's check in on you."],
  Developer: ["debug-and-fix", "Dev skills ready. Let's sharpen your code."],
  Finance: ["budget-builder", "Finance skills available. Get your money right."],
  Creative: ["content-machine", "Creative skills unlocked. Time to create."],
  Education: ["learn-anything", "Learning skills ready. What do you want to master?"],
};

const NEW_USER_SEQUENCE: Array<[string, string]> = [
  ["am-i-okay", "Welcome to SkillChain. Let's start with a check-in."],
  ["learn-anything", "Day 2. Let's learn something new together."],
  ["budget-builder", "Day 3. Let's look at your finances."],
  ["career-pivot", "Day 4. Let's think about your career."],
  ["side-hustle-launch", "Day 5. Ready to build something?"],
  ["weekly-ops", "Day 6. Let's plan your week."],
  ["content-machine", "Week 2. Time to create."],
  ["tech-confidence-builder", "Scared of AI? Let's fix that."],
  ["debug-and-fix", "Let's solve a real problem."],
  ["networking-blitz", "Time to connect with people."],
  ["big-decision", "Got a big choice? Let's think it through."],
  ["burnout-recovery", "Halfway through. Let's check your energy."],
  ["competitive-intel", "Know your landscape."],
  ["before-you-leap", "Almost two weeks in. What's your big move?"],
];

export class VelmaRecommender {
  private engine: GamificationEngine;
  private chainsDir: string;

  constructor(marketplaceDir?: string) {
    this.engine = new GamificationEngine();
    this.chainsDir = marketplaceDir
      ? join(marketplaceDir, "chains")
      : join(__dirname, "..", "..", "marketplace", "chains");
  }

  private loadChains(): Array<Record<string, unknown>> {
    if (!existsSync(this.chainsDir)) return [];
    return readdirSync(this.chainsDir)
      .filter(f => f.endsWith(".chain.json"))
      .map(f => {
        try { return JSON.parse(readFileSync(join(this.chainsDir, f), "utf-8")); }
        catch { return null; }
      })
      .filter(Boolean) as Array<Record<string, unknown>>;
  }

  whatNow(): VelmaRecommendation[] {
    const recs: VelmaRecommendation[] = [];
    const state = this.engine.getState();

    // 1. New user sequence (first 14 days)
    const totalRuns = state.total_skill_runs + state.total_chain_runs;
    if (totalRuns < NEW_USER_SEQUENCE.length) {
      const [chain, nudge] = NEW_USER_SEQUENCE[totalRuns];
      recs.push({
        chain_name: chain,
        reason: "new_user_progression",
        priority: 1,
        category: "onboarding",
        nudge,
      });
    }

    // 2. Streak maintenance
    const today = new Date().toISOString().slice(0, 10);
    if (state.streak_current > 0 && state.streak_last_date !== today) {
      recs.push({
        chain_name: "learn-anything",
        reason: `Don't break your ${state.streak_current}-day streak!`,
        priority: 1,
        category: "streak",
        nudge: `${state.streak_current}-day streak on the line. Run any skill to keep it alive.`,
      });
    }

    // 3. Time-of-day recommendation
    const hour = new Date().getHours();
    for (const [start, end, chain, nudge] of TIME_CHAINS) {
      if (hour >= start && hour < end) {
        recs.push({ chain_name: chain, reason: "time_of_day", priority: 3, category: "timing", nudge });
        break;
      }
    }

    // 4. Day-of-week recommendation
    const day = new Date().getDay();
    const dayRec = DAY_CHAINS[day];
    if (dayRec) {
      recs.push({ chain_name: dayRec[0], reason: "day_of_week", priority: 3, category: "timing", nudge: dayRec[1] });
    }

    // 5. Category gaps
    const discoveredCats = new Set(
      state.skills_discovered.map(sk => {
        for (const [cat, skills] of Object.entries(SKILL_CATEGORIES)) {
          if (skills.includes(sk)) return cat;
        }
        return null;
      }).filter(Boolean)
    );
    for (const [cat, [chain, nudge]] of Object.entries(CATEGORY_NUDGES)) {
      if (!discoveredCats.has(cat)) {
        recs.push({ chain_name: chain, reason: "category_gap", priority: 2, category: cat, nudge });
      }
    }

    // 6. Daily quest nudge
    if (state.daily_runs_date !== today || state.daily_runs_today.length === 0) {
      recs.push({
        chain_name: "learn-anything",
        reason: "daily_quests",
        priority: 2,
        category: "quests",
        nudge: "You have daily quests waiting. Run a skill to start earning bonus XP.",
      });
    }

    // Sort by priority, deduplicate by chain_name
    const seen = new Set<string>();
    return recs
      .sort((a, b) => a.priority - b.priority)
      .filter(r => {
        if (seen.has(r.chain_name)) return false;
        seen.add(r.chain_name);
        return true;
      })
      .slice(0, 5);
  }
}
